crates/sui-framework/packages/move-stdlib/sources/macros.move

```diff
@@ -145,3 +145,93 @@ public macro fun try_as_u128($x: _): Option<u128> {
     if (x > 0xFFFF_FFFF_FFFF_FFFF_FFFF_FFFF_FFFF_FFFF) option::none()
     else option::some(x as u128)
 }
+
+public macro fun uq_from_quotient<$T, $U>(
+    $numerator: $T,
+    $denominator: $T,
+    $max_t: $T,
+    $t_bits: u8,
+    $fractional_bits: u8,
+    $abort_denominator: _,
+    $abort_quotient_too_small: _,
+    $abort_quotient_too_large: _,
+): $T {
+    let numerator = $numerator;
+    let denominator = $denominator;
+    if (denominator == 0) $abort_denominator;
+
+    // Scale the numerator to have 64 fractional bits and the denominator to have 32 fractional
+    // bits, so that the quotient will have 32 fractional bits.
+    let scaled_numerator = numerator as $U << $t_bits;
+    let scaled_denominator = denominator as $U << $fractional_bits;
```

Comment:
It doesn't matter for the two concrete instantiations currently, but I think this doesn't work correctly if t_bits is not 2*fractional_bits. This should probably be:

```suggestion
    let scaled_numerator = numerator as $U << $t_bits;
    let scaled_denominator = denominator as $U << ($t_bits - $fractional_bits);
```